processing/src/main/java/org/apache/druid/segment/virtual/ListFilteredVirtualColumn.java

```diff
@@ -252,22 +256,172 @@ public BitmapFactory getBitmapFactory()
       return delegate.getBitmapFactory();
     }
 
+    @Override
+    public int getCardinality()
+    {
+      return idMapping.getValueCardinality();
+    }
+
+    @Override
+    public int getIndex(@Nullable String value)
+    {
+      return getReverseIndex(value);
+    }
+
     @Override
     public ImmutableBitmap getBitmap(int idx)
     {
       return delegate.getBitmap(idMapping.getReverseId(idx));
     }
 
     @Override
-    public int getCardinality()
+    public ImmutableBitmap getBitmapForValue(@Nullable String value)
     {
-      return idMapping.getValueCardinality();
+      if (getReverseIndex(value) < 0) {
+        return delegate.getBitmap(-1);
+      }
+      return delegate.getBitmapForValue(value);
     }
 
     @Override
-    public int getIndex(@Nullable String value)
+    public Iterable<ImmutableBitmap> getBitmapsInRange(
```

Comment:
For the "include" case, the implementations of these methods seem like they should be able to be a lot simpler.  The normal use case provides a much smaller list of things to show, it seems like we could quite rapidly filter through that list and then delegate to something that uses that list.

The deny-list case requires different logic.  But, then I wonder how important this actually is, or is the difference between the implementation that you have and this more direct implementation just a few milliseconds such that it doesn't actually matter...

Reply:
>For the "include" case, the implementations of these methods seem like they should be able to be a lot simpler. The normal use case provides a much smaller list of things to show, it seems like we could quite rapidly filter through that list and then delegate to something that uses that list.

I _think_ this is already doing this, the range iterated over here is the cardinality of the idLookup, not the dictionary of the delegate column, so it only considers ranges of the filtered values. There might be room for improvement, especially in the deny case, though I think it probably would need bigger changes to `ListFilteredVirtualColumn` and maybe best to do as a follow-up.